crates/core/component/auction/src/component/dutch_auction.rs

```diff
@@ -51,11 +64,183 @@ pub(crate) trait DutchAuctionManager: StateWrite {
+            // If the termination condition has been reached, we set the auction
+            // sequence to 1 (Closed).
+            new_dutch_auction.state.sequence = 1;
+        } else {
+            // Otherwise, we compute the next trigger height and generate a liquidity
+            // position for the new auction round.
+            let next_trigger = auction_trigger
+                .compute_next_trigger_height(trigger_height)
+                .expect("trigger data is validated")
+                .expect("the step count has not been reached");
+
+            // We compute the price parameters for the LP:
+            let (p, q) = compute_pq_at_step(&new_dutch_auction.description, step_index);
+
+            // Next, we want to construct an auction LP position and send it to the DEX.
+            // The nonce must be chosen so that the resulting position id is unique:
+            // `PositionManager::open_position` will reject duplicates.
+            //
+            // To do this, we keep track of our number of attempts at opening a position,
+            // and compute:
+            // position_nonce = H(auction_nonce || step_index || attempt_counter)
+            // until the resulting position id (based on the nonce) is unique and accepted
+            // by the DEX.
+            let mut attempt_counter = 0u64;
+            loop {
+                let pair = DirectedTradingPair::new(auction_input_id, auction_output_id);
+                let lp_reserves = Reserves {
+                    r1: new_dutch_auction.state.input_reserves,
+                    r2: Amount::zero(),
+                };
+                let auction_nonce = new_dutch_auction.description.nonce;
+                let full_hash = blake2b_simd::Params::default()
+                    .personal(b"penum-DA-nonce")
+                    .to_state()
+                    .update(&auction_nonce)
+                    .update(&step_index.to_le_bytes())
+                    .update(&attempt_counter.to_le_bytes())
+                    .finalize();
+                let mut tough_nonce = [0u8; 32];
+                tough_nonce[0..32].copy_from_slice(&full_hash.as_bytes()[0..32]);
+
+                let mut lp = Position::new_with_nonce(tough_nonce, pair, 0u32, p, q, lp_reserves);
+                lp.close_on_fill = true;
+
+                let id = lp.id();
+                new_dutch_auction.state.current_position = Some(id);
+
+                match self.open_position(lp).await {
+                    Ok(_) => break,
+                    Err(e) => {
+                        // TODO(erwan): this coarsely assumes that all errors
+                        // are id collisions which obviously does not hold.
+                        // for now it's impossible to do finer grained control flow.
+                        tracing::error!(?e, attempt_counter, ?id, "failed to open position");
+                        attempt_counter += 1;
+                    }
+                }
```

Comment:
Why don't we just check that the position ID is unused before opening it? Then we can be precise about our control flow, at the cost of duplicating a single state read (on average one time).

Reply:
Note: we'll need to make sure that the control flow doesn't leak value, if we error creating a position for whatever reason we should try to leave the state consistent (put value back into the auction object or something), or panic